java/lang/StringBuilder/StringBuilder.cpp

```diff
@@ -0,0 +1,868 @@
+    memmove(newPositionOfTailPart, oldPositionOfTailPart, (size_t)memorySizeOfTailPart);
+
+    string insertPosition = this->original + start;
+    int memorySizeForTarget = lengthOfTarget * sizeof(char);
+    memcpy(insertPosition, target, (size_t)memorySizeForTarget);
+
+    this->currentLength = newLength;
+    return *this;
+}
+
+StringBuilder StringBuilder::reverse() {
+    boolean hasSurrogates = false;
+    int index;
+    int oppositeIndex;
+    int stopIndex = this->currentLength / 2;
+    char temp;
+    for (index = 0; index < stopIndex; index++) {
+        oppositeIndex = (this->currentLength - 1) - index;
+        temp = this->original[index];
+        this->original[index] = this->original[oppositeIndex];
+        this->original[oppositeIndex] = temp;
+        if (Character::isSurrogate((this->original[index])) || Character::isSurrogate(this->original[oppositeIndex])) {
+            hasSurrogates = true;
+        }
+    }
+    if (hasSurrogates) {
+        this->reverseAllValidSurrogatePairs();
+    }
+    return *this;
+}
+
+void StringBuilder::setCharAt(int index, char target) {
+    if (index < 0 || index >= this->currentLength) {
+        throw StringIndexOutOfBoundsException(index);
+    }
+
+    this->original[index] = target;
+}
+
+void StringBuilder::setCharAt(int index, const Character &target) {
+    Character *pointerToTarget = const_cast<Character *>(&target);
+    this->setCharAt(index, pointerToTarget->charValue());
+}
+
+void StringBuilder::setLength(int newLength) {
+    if (newLength < 0) {
+        throw StringIndexOutOfBoundsException(newLength);
+    }
+
+    this->ensureCapacity(newLength);
+    if (this->currentLength < newLength) {
+        int index;
+        for (index = this->currentLength; index < newLength; index++) {
+            this->original[index] = '\0';
+        }
+    }
+    this->currentLength = newLength;
+}
+
+String StringBuilder::substring(int start) const {
+    return this->substring(start, this->currentLength);
+}
+
+String StringBuilder::substring(int start, int end) const {
+    if (start < 0 || start > this->currentLength) {
+        throw StringIndexOutOfBoundsException(start);
+    }
+    if (end < 0 || end > this->currentLength) {
+        throw StringIndexOutOfBoundsException(end);
+    }
+    if (start > end) {
+        throw StringIndexOutOfBoundsException(end - start);
+    }
+
+    int numberOfBytesForSubString = (end - start + 1) * sizeof(char);
+    string copyOfSubString = (string)malloc((size_t)numberOfBytesForSubString);
+    int indexOfOriginal;
+    int indexOfSubString = 0;
+    for (indexOfOriginal = start; indexOfOriginal < end;indexOfOriginal++) {
+        copyOfSubString[indexOfSubString] = this->original[indexOfOriginal];
+        indexOfSubString = indexOfSubString + 1;
+    }
+    copyOfSubString[indexOfSubString] = '\0';
+    String result(copyOfSubString);
+    free(copyOfSubString);
+    return result;
+}
+
+String StringBuilder::toString() const {
```

Comment:
I use a C char array without null terminator to save character sequence for StringBuilder (for convenient), so I must copy that character sequence and add a null terminator (if I use String constructor, it will fail).